json_visitor.go

```diff
@@ -0,0 +1,278 @@
+package raymond
+
+import (
+	"encoding/json"
+	"github.com/mailgun/raymond/v2/ast"
+	"strings"
+)
+
+type list []interface{}
+
+func (l *list) Add(item interface{}) {
+	*l = append(*l, item)
+}
+
+func newList(item interface{}) *list {
+	l := new(list)
+	l.Add(item)
+	return l
+}
+
+type JSONVisitor struct {
+	JSON map[string]interface{}
+	ctx  *handlebarsContext
+}
+
+func newJSONVisitor() *JSONVisitor {
+	j := map[string]interface{}{}
+	v := &JSONVisitor{JSON: j, ctx: newHandlebarsContext()}
+	return v
+}
+
+func ToJSON(node ast.Node) string {
+	visitor := newJSONVisitor()
+	node.Accept(visitor)
+	b, _ := json.Marshal(visitor.JSON)
+	return string(b)
+}
+
+func (v *JSONVisitor) VisitProgram(node *ast.Program) interface{} {
+	for _, n := range node.Body {
+		n.Accept(v)
+	}
+	return v.JSON
+}
+
+func (v *JSONVisitor) VisitMustache(node *ast.MustacheStatement) interface{} {
+
+	node.Expression.Accept(v)
+
+	return nil
+}
+
+func (v *JSONVisitor) VisitBlock(node *ast.BlockStatement) interface{} {
+	var action string
+	fp := node.Expression.FieldPath()
+	if fp != nil {
+		action = fp.Parts[0]
+	}
+	if action == "with" || action == "each" {
+		blockParamsPath := make([]string, 0)
+		blockParams := make([]string, 0)
+		for _, params := range node.Expression.Params {
+			// Extract block params from nested nodes.
+			if pe, ok := params.(*ast.PathExpression); ok {
+				blockParamsPath = append(blockParamsPath, pe.Parts...)
+			}
+		}
+		if node.Program != nil {
+			if len(node.Program.BlockParams) > 0 {
+				blockParams = append(node.Program.BlockParams)
+			}
+		}
+		if action == "each" {
+			blockParamsPath = append(blockParamsPath, "[0]")
+		}
+		if len(blockParams) > 0 {
+			v.ctx.AddMemberContext(strings.Join(blockParamsPath, "."), strings.Join(blockParams, "."))
+		} else {
+			v.ctx.AddMemberContext(strings.Join(blockParamsPath, "."), "")
+		}
+		if node.Program != nil {
+			node.Program.Accept(v)
+		}
+		if node.Inverse != nil {
+			node.Inverse.Accept(v)
+		}
+		v.ctx.MoveUpContext()
+	} else {
+		for _, param := range node.Expression.Params {
+			param.Accept(v)
+		}
+		if node.Program != nil {
+			node.Program.Accept(v)
+		}
+		if node.Inverse != nil {
+			node.Inverse.Accept(v)
+		}
+	}
+	return nil
+}
+
+func (v *JSONVisitor) VisitPartial(node *ast.PartialStatement) interface{} {
+
+	return nil
+}
+
+func (v *JSONVisitor) VisitContent(node *ast.ContentStatement) interface{} {
+
+	return nil
+}
+
+func (v *JSONVisitor) VisitComment(node *ast.CommentStatement) interface{} {
+
+	return nil
+}
+
+func (v *JSONVisitor) VisitExpression(node *ast.Expression) interface{} {
+	var action string
+	fp := node.FieldPath()
+	if fp != nil {
+		if len(fp.Parts) > 0 {
+			action = fp.Parts[0]
+			if action == "lookup" {
+				path, ok := node.Params[0].(*ast.PathExpression)
```

Comment:
Similar concern. Can `node.params` be emtpy? Would the `ok` catch that too or just catch the type conversion?

Reply:
Fixed